File: python/src/threads/tools/lsp.py
"""The lsp tool: read_only, run in the sandbox against the image's language
servers. Each call uploads the driver (`lsp_driver.py`), which starts the server, asks one
question and exits. A server that is missing or doesn't answer is `unavailable`, never an empty
success.

ponytail: a server per call; keep one running per language if indexing time hurts.
"""

import json
import posixpath
from collections.abc import Mapping, Sequence
from pathlib import Path
from typing import Final

from pydantic import JsonValue, TypeAdapter, ValidationError
from pydantic.experimental.missing_sentinel import MISSING

from threads._generated.tools_v1 import LspInput
from threads.loop.tools import Dispatched, NotSent, Output
from threads.result import Err
from threads.tools import files
from threads.tools.control import Control

LANGUAGES: Final[Mapping[str, tuple[str, ...]]] = {
    "python": ("pyright-langserver", "--stdio"),
    "typescript": ("typescript-language-server", "--stdio"),
    "go": ("gopls",),
    "rust": ("rust-analyzer",),
}
"""Each declarable language and the server command its image must have."""
_IDS: Final = {
    ".py": ("python", "python"),
    ".pyi": ("python", "python"),
    ".ts": ("typescript", "typescript"),
    ".tsx": ("typescript", "typescriptreact"),
    ".js": ("typescript", "javascript"),
    ".jsx": ("typescript", "javascriptreact"),
    ".mjs": ("typescript", "javascript"),
    ".cjs": ("typescript", "javascript"),
    ".go": ("go", "go"),
    ".rs": ("rust", "rust"),
}
"""A file extension's declared language and its LSP languageId."""
DRIVER: Final = "/workspace/.threads/lsp_driver.py"
_TIMEOUT_MS: Final = 120_000
_ANSWER: Final = TypeAdapter(dict[str, JsonValue])
_SEVERITY: Final = {1: "error", 2: "warning", 3: "info", 4: "hint"}


async def run(  # noqa: PLR0911 - one early answer per failure
    tools: Control,
    args: LspInput,
    key: str,
    servers: Mapping[str, Sequence[str]],
) -> Dispatched:
    """`servers`: the declared languages' server commands."""
    path = files.absolute(args.path)
    language, language_id = _IDS.get(posixpath.splitext(path)[1], ("", ""))
    if language not in servers:
        return Output(f"unavailable: no language server is declared for {args.path}", True)
    positional = args.operation in ("definition", "references", "hover")
    if positional and (args.line is MISSING or args.character is MISSING):
        return Output(f"{args.operation} needs line and character", True)
    session = await tools.session()
    if session is None:
        return NotSent()
    driver = (Path(__file__).parent / "lsp_driver.py").read_bytes()
    up = await session.upload(DRIVER, driver, tools.context)
    if isinstance(up, Err):
        return Output(f"unavailable: {up.error.code}: {up.error.message}", True)
    argv = [
        "python3",
        DRIVER,
        json.dumps(list(servers[language])),
        "/workspace",
        path,
        language_id,
        args.operation,
    ]
    if args.line is not MISSING and args.character is not MISSING:
        argv += [str(args.line), str(args.character)]
    ran = await tools.command(argv, key, _TIMEOUT_MS)
    if isinstance(ran, Err):
        return ran.error
    try:
        answer = _ANSWER.validate_json(ran.value.stdout.strip().rsplit("\n", 1)[-1])
    except ValidationError:
        return Output(f"unavailable: the lsp driver failed: {ran.value.stderr.strip()}", True)
    if "unavailable" in answer or answer.get("result") is None:
        why = answer.get("unavailable") or f"the server gave no {args.operation}"
        return Output(f"unavailable: {why}", True)
    return Output(render(args.operation, answer["result"]))


def _start(span: JsonValue) -> str:
    """A range's start as line:character, 1-based, or empty."""
    start = span.get("start") if isinstance(span, dict) else None
    line = start.get("line") if isinstance(start, dict) else None
    char = start.get("character") if isinstance(start, dict) else None
    if isinstance(line, int) and isinstance(char, int):
        return f"{line + 1}:{char + 1}"
    return ""


def _where(value: JsonValue) -> str:
    """A Location or LocationLink as path:line:character."""
    if not isinstance(value, dict):
        return str(value)
    uri = str(value.get("uri", value.get("targetUri"))).removeprefix("file://")
    at = _start(value.get("range", value.get("targetSelectionRange")))
    return f"{uri}:{at}" if at else uri


def _hover(contents: JsonValue) -> str:
    if isinstance(contents, list):
        return "\n\n".join(_hover(c) for c in contents)
    if isinstance(contents, dict):
        return str(contents.get("value", ""))
    return str(contents)


def _symbols(items: Sequence[JsonValue], depth: int = 0) -> list[str]:
    """DocumentSymbol trees or SymbolInformation lists."""
    out: list[str] = []
    for item in items:
        if not isinstance(item, dict):
            continue
        location = item.get("location")
        span = location.get("range") if isinstance(location, dict) else item.get("range")
        at = _start(span) or "?"
        out.append(f"{'  ' * depth}{item.get('name')} (kind {item.get('kind')}) at {at}")
        children = item.get("children")
        if isinstance(children, list):
            out += _symbols(children, depth + 1)
    return out


def _diagnostic(item: JsonValue) -> str:
    if not isinstance(item, dict):
        return str(item)
    severity = item.get("severity")
    label = _SEVERITY.get(severity, "diagnostic") if isinstance(severity, int) else "diagnostic"
    return f"{_start(item.get('range'))} {label}: {item.get('message')}"


def render(operation: str, result: JsonValue) -> str:
    """The server's answer as text, positions 1-based."""
    items: Sequence[JsonValue] = result if isinstance(result, list) else [result]
    match operation:
        case "diagnostics":
            return "\n".join(_diagnostic(d) for d in items) or "no diagnostics"
        case "definition" | "references":
            return "\n".join(_where(i) for i in items) or f"no {operation}"
        case "hover":
            text = _hover(result.get("contents") if isinstance(result, dict) else result)
            return text or "no hover information"
        case _:
            return "\n".join(_symbols(items)) or "no symbols"

File: typescript/packages/core/src/tools/lsp.ts
import { posix } from "node:path";
import { z } from "zod";
import { ConfigError } from "../agent/errors";
import { assertNever } from "../assert-never";
import type { ToolRun } from "../loop/types";
import { execute, toolRunOf } from "../sandbox/exec";
import { type Builtin, builtin, done, sessionOf } from "./builtin";
import { LSP_DRIVER } from "./lsp-driver";
import { LspInput } from "./sandbox-inputs";

// lsp: read_only, run in the sandbox against the language servers the image
// has, one per declared language. A server that is missing or not ready answers `unavailable`,
// never an empty success.

type Server = {
  readonly command: readonly string[];
  readonly extensions: Readonly<Record<string, string>>;
};
type Operation = z.infer<typeof LspInput>["operation"];

/** The preset servers per language: the argv and the LSP languageId per file extension. */
export const LSP_SERVERS: Readonly<Record<string, Server>> = {
  typescript: {
    command: ["typescript-language-server", "--stdio"],
    extensions: {
      ".ts": "typescript",
      ".tsx": "typescriptreact",
      ".js": "javascript",
      ".jsx": "javascriptreact",
    },
  },
  python: {
    command: ["pyright-langserver", "--stdio"],
    extensions: { ".py": "python" },
  },
  go: { command: ["gopls"], extensions: { ".go": "go" } },
  rust: { command: ["rust-analyzer"], extensions: { ".rs": "rust" } },
};

const READY_S = 30;
const WAIT_S = 20;
const TIMEOUT_MS = (READY_S + WAIT_S + 10) * 1000;

// The driver's one line and the server's answers are sandbox output: parsed at the boundary.
const Answer = z.union([
  z.object({ ok: z.json() }),
  z.object({ unavailable: z.string() }),
]);
const Position = z.object({ line: z.int(), character: z.int() });
const Range = z.object({ start: Position, end: Position });
const Location = z.object({ uri: z.string(), range: Range });
const Link = z.object({ targetUri: z.string(), targetSelectionRange: Range });
const Diagnostic = z.object({
  range: Range,
  severity: z.int().optional(),
  message: z.string(),
});
type DocSymbol = {
  readonly name: string;
  readonly kind: number;
  readonly range?: z.infer<typeof Range> | undefined;
  readonly location?: z.infer<typeof Location> | undefined;
  readonly children?: readonly DocSymbol[] | undefined;
};
const DocSymbol: z.ZodType<DocSymbol> = z.lazy(() =>
  z.object({
    name: z.string(),
    kind: z.int(),
    range: Range.optional(),
    location: Location.optional(),
    children: z.array(DocSymbol).optional(),
  }),
);
const Marked = z.union([z.string(), z.object({ value: z.string() })]);
const Hover = z.object({ contents: z.union([Marked, z.array(Marked)]) });

const SEVERITY = ["", "error", "warning", "info", "hint"];
const at = (uri: string, p: z.infer<typeof Position>): string =>
  `${uri.replace(/^file:\/\//, "")}:${p.line + 1}:${p.character + 1}`;
const listed = (found: readonly string[], none: string): string =>
  found.length === 0 ? none : found.join("\n");

function outline(list: readonly DocSymbol[], depth = 0): string[] {
  return list.flatMap((s) => {
    const start = s.range?.start ?? s.location?.range.start;
    const line = start === undefined ? "" : ` line ${start.line + 1}`;
    return [
      `${"  ".repeat(depth)}${s.name} (kind ${s.kind})${line}`,
      ...outline(s.children ?? [], depth + 1),
    ];
  });
}

function diagnostics(uri: string, result: unknown): string | undefined {
  const d = z.array(Diagnostic).safeParse(result);
  if (!d.success) return undefined;
  const line = (x: z.infer<typeof Diagnostic>): string =>
    `${at(uri, x.range.start)} ${SEVERITY[x.severity ?? 1] ?? "info"}: ${x.message}`;
  return listed(d.data.map(line), "no diagnostics");
}

function locations(op: string, result: unknown): string | undefined {
  const one = z.union([Location, Link]);
  const got = z.union([z.null(), one, z.array(one)]).safeParse(result);
  if (!got.success) return undefined;
  const all = got.data === null ? [] : [got.data].flat();
  const where = (l: z.infer<typeof one>): string =>
    "uri" in l
      ? at(l.uri, l.range.start)
      : at(l.targetUri, l.targetSelectionRange.start);
  return listed(all.map(where), `no ${op} found`);
}

function hover(result: unknown): string | undefined {
  const h = z.union([z.null(), Hover]).safeParse(result);
  if (!h.success) return undefined;
  if (h.data === null) return "no hover information";
  return [h.data.contents]
    .flat()
    .map((p) => (typeof p === "string" ? p : p.value))
    .join("\n");
}

/** The model-visible text of an LSP answer, or undefined when it isn't the expected shape. */
export function lspText(
  operation: Operation,
  uri: string,
  result: unknown,
): string | undefined {
  switch (operation) {
    case "diagnostics":
      return diagnostics(uri, result);
    case "definition":
    case "references":
      return locations(operation, result);
    case "hover":
      return hover(result);
    case "symbols": {
      const s = z.union([z.null(), z.array(DocSymbol)]).safeParse(result);
      return s.success
        ? listed(outline(s.data ?? []), "no symbols")
        : undefined;
    }
    default:
      return assertNever(operation);
  }
}

/** The driver's printed answer as the tool's result. */
export function answered(
  operation: Operation,
  path: string,
  stdout: string,
): ToolRun {
  let raw: unknown;
  try {
    raw = JSON.parse(stdout.trim().split("\n").at(-1) ?? "");
  } catch {
    return done("unavailable: the lsp client printed no answer", true);
  }
  const answer = Answer.safeParse(raw);
  if (!answer.success)
    return done("unavailable: the lsp client answered malformed output", true);
  if ("unavailable" in answer.data)
    return done(`unavailable: ${answer.data.unavailable}`, true);
  const shown = lspText(operation, `file://${path}`, answer.data.ok);
  return shown === undefined
    ? done(`unavailable: an unexpected ${operation} answer`, true)
    : done(shown);
}

function servers(languages: readonly string[]): readonly Server[] {
  return languages.map((name) => {
    const server = LSP_SERVERS[name];
    if (server === undefined)
      throw new ConfigError(
        "unknown_preset",
        `lsp: no language server preset ${name}; known: ${Object.keys(LSP_SERVERS).join(", ")}`,
      );
    return server;
  });
}

/** lsp for the declared languages; an unknown one is a setup error. */
export function lsp(languages: readonly string[]): Builtin {
  return lspWith(servers(languages));
}

/** lsp over these servers. */
export function lspWith(declared: readonly Server[]): Builtin {
  return builtin({
    name: "lsp",
    input: LspInput,
    effect: "read_only",
    run: async (input, ctx, env) => {
      const path = posix.resolve("/workspace", input.path);
      const ext = posix.extname(path);
      const server = declared.find((s) => s.extensions[ext] !== undefined);
      if (server === undefined)
        return done(
          `unavailable: no declared language server handles ${path}`,
          true,
        );
      const positioned =
        input.operation !== "diagnostics" && input.operation !== "symbols";
      if (
        positioned &&
        (input.line === undefined || input.character === undefined)
      )
        return done(`${input.operation} needs line and character`, true);
      const session = await sessionOf(env);
      if (!session.ok) return session.error;
      const args = {
        command: server.command,
        root: "/workspace",
        path,
        language_id: server.extensions[ext],
        operation: input.operation,
        line: input.line ?? 1,
        character: input.character ?? 1,
        ready_s: READY_S,
        wait_s: WAIT_S,
      };
      const ran = await execute(
        session.value,
        ["python3", "-c", LSP_DRIVER, JSON.stringify(args)],
        env.context,
        { env: {}, timeoutMs: TIMEOUT_MS, processKey: ctx.effectKey },
        env.artifacts,
        1 << 20,
      );
      if (!ran.ok) return toolRunOf(ran);
      if (ran.value.exit_code !== 0)
        return done(
          `unavailable: the lsp client failed: ${ran.value.stderr}`,
          true,
        );
      return answered(input.operation, path, ran.value.stdout);
    },
  });
}
